### PR TITLE
Model evaluation

### PR DESCRIPTION
**Pull request by tarek-nawara**
Separated the model evaluation in a separate class so we can reuse the evaluation code.
issue #20

### CODE REVIEW

File: src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    EsSpark.saveToEs(evaluation, "training-analysis")
+  }
+
+  /**
+    * Evaluate all the testing data.
+    *
+    * @param model target model for evaluation
+    * @return rdd of `EvaluatedTrainingTweet` instances.
+    */
+  private def evaluateData(model: GradientBoostingModel#GenericModel): RDD[EvaluatedTrainingTweet] = {
```
Comment (mohamedramadan2515) [inline src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala]: We couldn't simply load all data we need to separate training error from the testing error

File: src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala [hunk 2]
```diff
@@ -0,0 +1,116 @@
+package edu.twitter.evaluation
```
Comment (mohamedramadan2515) [inline src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala]: I believe this class constructor has to take the same training and testing data used to train the model!!!?

File: src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala (hunk 1 above)
Comment (tarek-nawara) [reply src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala]: Good point, I was going to add a method just for that. we will provide it with training and testing data

File: src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala (hunk 2 above)
Comment (tarek-nawara) [reply src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala]: I agree. As I mentioned I will add the method for that but we need also the ability to evaluate a given model with labeled data in order to use something like `textblob` in the future for evaluation.

Comment (mohamedramadan2515) [reply src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala]: we can make a method that takes data as a parameter

Comment (tarek-nawara) [reply src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala]: We can consider something like that. but I guess from encapsulation perspective we can leave the implemented `evaluate` method.

Comment (MohamedElhawaty) [reply src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala]: We need to distinguish between two different times where the evaluation is required:
1) Evaluating the model after it's build using the test data.
2) Evaluating a given model using a specific labeled data set from ( textblob or google cloud api) in order to compare it with other models.

The first part is more related to the building process and in my opinion, it should be moved to the ModelBuilder. Fortunately, as we are mainly concerned with the accuracy for this task, each of the two models that we are building has its already implemented method for the evaluation:
- evaluateEachIteration() for the Gradient Boosted Trees Model
- evaluate() for the Neural Network Model

This way, this class would be responsible for the second task only.

Comment (tarek-nawara) [reply src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala]: Totally agree, this idea will be much cleaner in the implementation :+1:

File: src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+  * @param sc spark context.
+  */
+class ModelEvaluator(sc: SparkContext) {
+  /**
+    * Given a model will evaluate it based on
+    * training data and log evaluation analysis.
+    *
+    * @param model target model for evaluation.
+    */
+  def evaluate(model: GenericModel): Unit = {
```
Comment (mohamedramadan2515) [inline src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala]: Sorry, what is the use of this method now?

File: src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+  }
+
+  /**
+    * Given a model will evaluate it based on
+    * training data, log evaluation analysis and
+    * persist the labeled records in `Elasticsearch`
+    *
+    * @param model target model for evaluation
+    */
+  def evaluateAndPersist(model: GenericModel): Unit = {
```
Comment (mohamedramadan2515) [inline src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala]: I think this also has to be related to training and testing data and has to save only misclassified data, Can we also add label is this misclassified tweet from training or testing?

File: src/test/scala/edu/twitter/evaluation/ModelEvaluatorTest.scala [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import org.apache.spark.{SparkConf, SparkContext}
+
+object ModelEvaluatorTest {
+  def main(args: Array[String]): Unit = {
+    val conf = new SparkConf().setMaster("local[*]").setAppName("Twitter")
+    val sc = new SparkContext(conf)
+
+    val model = new GradientBoostingBuilder(sc).build()
+    val modelEvaluator = new ModelEvaluator(sc)
+    modelEvaluator.evaluate(model)
```
Comment (mohamedramadan2515) [inline src/test/scala/edu/twitter/evaluation/ModelEvaluatorTest.scala]: Can you also test the evaluation with testing and training data?

File: src/main/scala/edu/twitter/model/impl/GradientBoostingBuilder.scala [hunk 6]
```diff
@@ -23,6 +26,9 @@ class GradientBoostingBuilder(sc: SparkContext) extends GenericModelBuilder {
     * @return GenericModel
     */
   def build(): GenericModel = {
+    implicit def toRecords(data: RDD[(String, LabeledPoint)]): RDD[Record] = data.map { case (msg, p) => Record(msg, p.label) }
```
Comment (mohamedramadan2515) [inline src/main/scala/edu/twitter/model/impl/GradientBoostingBuilder.scala]: Sorry I couldn't find where we use this?

File: src/main/scala/edu/twitter/model/impl/GradientBoostingBuilder.scala [hunk 7]
```diff
@@ -23,6 +26,9 @@ class GradientBoostingBuilder(sc: SparkContext) extends GenericModelBuilder {
...
+    implicit def toLabeledPoint(data: RDD[(String, LabeledPoint)]): RDD[LabeledPoint] = data.map(_._2)
```
Comment (mohamedramadan2515) [inline src/main/scala/edu/twitter/model/impl/GradientBoostingBuilder.scala]: Sorry I couldn't find where we use this?

File: src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala (hunk 3 above)
Comment (tarek-nawara) [reply src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala]: This method will evaluate a given `GenericModel` using data stored in `resources` we plan to use it to evaluate the models using data labeled by `textblob`

File: src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala (hunk 4 above)
Comment (tarek-nawara) [reply src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala]: I don't think we are not interested in visualizing training and testing errors. we should only compare the model's prediction to a labeled data coming from other model like Google cloud's model or `textblob`.

File: src/main/scala/edu/twitter/model/impl/GradientBoostingBuilder.scala (hunk 7 above)
Comment (tarek-nawara) [reply src/main/scala/edu/twitter/model/impl/GradientBoostingBuilder.scala]: they are being used in converting `RDD[(String, LabeledPoing)]` to `RDD[LabeledPoint]` and RDD[Record]`. the scala compiler will invoke the method when it needs the target type. you can find more about `implicits` in scala [here](https://docs.scala-lang.org/tour/implicit-conversions.html).

File: src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala (hunk 4 above)
Comment (mohamedramadan2515) [reply src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala]: I believe we are interested in visualizing training and testing errors!!

File: src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala (hunk 3 above)
Comment (mohamedramadan2515) [reply src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala]: So why tweetsLoader don't take dataPath as a parameter?

File: src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    val tweetsLoader = new TweetsLoader(sc)
+    val evaluation = evaluateData(model, tweetsLoader.getTweetsDataSet())
+    performEvaluationAnalysis(evaluation)
+    EsSpark.saveToEs(evaluation, "training-analysis")
```
Comment (sherif98) [inline src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala]: can you change the index name of the elastic search to be something like "index name/ mapping name"
like "training/analysis".
it will help if we want to add a schema to the index later.

File: src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala (hunk 3 above)
Comment (tarek-nawara) [reply src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala]: Good point, I will add it :+1:

File: src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala (hunk 4 above)
Comment (tarek-nawara) [reply src/main/scala/edu/twitter/model/evaluation/ModelEvaluator.scala]: Ok, will add it to the evaluation.

File: src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala (hunk 2 above)
Comment (MohamedElhawaty) [reply src/main/scala/edu/twitter/evaluation/ModelEvaluator.scala]: I don't know why but evaluateEachIteration() doesn't work as I was expecting, I will check it again later. 

For now, could you please consider using MulticlassMetrics? I have tried it and it looks good.  
https://spark.apache.org/docs/2.2.0/mllib-evaluation-metrics.html#tab_scala_1
